rmsgwmon segfault on Bookworm

Hi,

I've recently relocated my RMS gateway over to a RPi4 running Bookworm.  I rebuilt rmsgw 2.5.1 using sources from https://github.com/nwdigitalradio/rmsgw .  When I try to run rmsgwmon, it immediately crashes with a "Segmentation fault" error.

Using print statements, I've managed to isolate this segfault to a call to wattroff in function aciStatus() on line 165 of rmsgwmon.c:

          `wattroff(win, COLOR_PAIR(SCR_COLOR_ACI_IDLE) | A_BOLD);`

I am using version 6.4-4 of the libncurses6 package, which was the default with Bookworm.  On my older system, I was running Debian Stretch with the libncurses5 package and rmsgwmon appeared to work.

  Thanks
  Mike

Here is the result of a debug session with gdb: 
```gdb
# gdb ./rmsgwmon/rmsgwmon 
GNU gdb (Debian 13.1-3) 13.1
Copyright (C) 2023 Free Software Foundation, Inc.
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.
Type "show copying" and "show warranty" for details.
This GDB was configured as "aarch64-linux-gnu".
Type "show configuration" for configuration details.
For bug reporting instructions, please see:
<https://www.gnu.org/software/gdb/bugs/>.
Find the GDB manual and other documentation resources online at:
    <http://www.gnu.org/software/gdb/documentation/>.

For help, type "help".
Type "apropos word" to search for commands related to "word"...
Reading symbols from ./rmsgwmon/rmsgwmon...
(gdb) tty /dev/null
(gdb) break rmsgwmon.c:162
Breakpoint 1 at 0x2e84: file ./rmsgwmon/rmsgwmon.c, line 162.
(gdb) run
Starting program: /usr/local/src/rmsgw/rmsgwmon/rmsgwmon 
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/lib/aarch64-linux-gnu/libthread_db.so.1".

Breakpoint 1, aciStatus (win=0x555559caf0) at ./rmsgwmon/rmsgwmon.c:162
162		  wprintw(win, "%s (%d)",
(gdb) step
map_aci_state (s=ACI_IDLE) at ./lib/statblock.c:126
126	char *map_aci_state(ACISTATE s) {
(gdb) step
128	     return(aci_state_map[s]);
(gdb) print aci_state_map[s]
$1 = 0x5555555d68 "Idle"
(gdb) stepi
0x0000005555554058	128	     return(aci_state_map[s]);
(gdb) stepi
0x000000555555405c	129	}
(gdb) stepi
0x0000005555554060	129	}
(gdb) stepi
0x0000005555552e90 in aciStatus (win=0x555559caf0) at ./rmsgwmon/rmsgwmon.c:162
162		  wprintw(win, "%s (%d)",
(gdb) stepi
0x0000005555552e94	162		  wprintw(win, "%s (%d)",
(gdb) stepi
0x0000005555552e98	162		  wprintw(win, "%s (%d)",
(gdb) stepi
0x0000005555552e9c	162		  wprintw(win, "%s (%d)",
(gdb) stepi
0x0000005555552ea0	162		  wprintw(win, "%s (%d)",
(gdb) stepi
0x0000005555552ea4	162		  wprintw(win, "%s (%d)",
(gdb) stepi
0x0000005555552ea8	162		  wprintw(win, "%s (%d)",
(gdb) x/s $x1
0x5555555208:	"%s (%d)"
(gdb) x/s $x2
0x55555d68:	<error: Cannot access memory at address 0x55555d68>
```

After a careful inspection of a build log, I found out the following warnings:
```
rmsgwmon.c: In function ‘aciStatus’:
rmsgwmon.c:163:19: warning: implicit declaration of function ‘map_aci_state’; did you mean ‘sbset_aci_state’? [-Wimplicit-function-declaration]
  163 |                   map_aci_state(rms_stat->aci.aci_state),
      |                   ^~~~~~~~~~~~~
      |                   sbset_aci_state
rmsgwmon.c:162:26: warning: format ‘%s’ expects argument of type ‘char *’, but argument 3 has type ‘int’ [-Wformat=]
  162 |           wprintw(win, "%s (%d)",
      |                         ~^
      |                          |
      |                          char *
      |                         %d
  163 |                   map_aci_state(rms_stat->aci.aci_state),
      |                   ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
      |                   |
      |                   int
rmsgwmon.c:168:26: warning: format ‘%s’ expects argument of type ‘char *’, but argument 3 has type ‘int’ [-Wformat=]
  168 |           wprintw(win, "%s (%d) (%6d)",
      |                         ~^
      |                          |
      |                          char *
      |                         %d
  169 |                   map_aci_state(rms_stat->aci.aci_state),
      |                   ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
      |                   |
      |                   int
```

Basically, because of an implicit declaration of `map_aci_state()`, the compiler implied the return value to be a 32 bit signed integer instead of a pointer, hence explaining the pointer aliasing. It didn't cause issues on 32 bits systems, due to matching pointer sizes.

--- a/Makefile.am
+++ b/Makefile.am
@@ -12,7 +12,7 @@ do_subst:
 
 CLEANFILES = $(pathheader)
 
-SUBDIRS = include lib rmsgw rmsgw_aci rmsgwmon rmschanstat devutils man contrib doc etc admin
+SUBDIRS = include lib rmsgw rmsgw_aci rmsgwmon rmschanstat rmsfinddev devutils man contrib doc etc admin
 
 ACLOCAL_AMFLAGS = -I m4
 

--- a/configure.ac
+++ b/configure.ac
@@ -50,6 +50,7 @@ AC_CONFIG_FILES([Makefile \
 	rmsgw_aci/Makefile \
 	rmsgwmon/Makefile \
 	rmschanstat/Makefile \
+	rmsfinddev/Makefile \
 	devutils/Makefile \
 	man/Makefile \
 	contrib/Makefile \

--- a/debian/control
+++ b/debian/control
@@ -2,7 +2,7 @@ Source: rmsgw
 Section: hamradio
 Priority: optional
 Maintainer: Lee Woldanski <ve7fet@tparc.org>
-Build-Depends: debhelper (>= 10), autotools-dev, libncurses5-dev, libxml2-dev, libc6 (>= 2.7-1~)
+Build-Depends: debhelper (>= 10), autotools-dev, libax25-dev, libncurses5-dev, libxml2-dev, libc6 (>= 2.7-1~)
 Standards-Version: 4.1.2
 Homepage: https://github.com/nwdigitalradio/rmsgw
 

--- a/include/rmslib.h
+++ b/include/rmslib.h
@@ -247,6 +247,9 @@ extern int sendDataGram(char *host, char *port, unsigned char* msg, int msg_len)
 /*
  * status block function prototypes
  */
+extern const char *map_aci_state(ACISTATE s);
+extern const char *map_gw_state(GWSTATE s);
+extern const char *map_gw_comm(GWSTATE s);
 extern int set_shm_debug(void);
 extern int get_cms_slot(rms_status *p, char *cms_host);
 extern int statblock_attach(rms_status **addr);

--- a/lib/statblock.c
+++ b/lib/statblock.c
@@ -48,7 +48,7 @@ static char svnid[] = "$Id: statblock.c 109 2009-08-10 08:31:14Z eckertb $";
  */
 extern int shm_debug;
 
-static char *aci_state_map[] = {
+static const char *aci_state_map[] = {
      "Unused",
      "Running",
      "Idle",
@@ -62,7 +62,7 @@ static char *aci_state_map[] = {
      NULL
 };
 
-static char *gw_state_map[] = {
+static const char *gw_state_map[] = {
      "Unused",
      "Running",
      "Idle",
@@ -79,7 +79,7 @@ static char *gw_state_map[] = {
      NULL
 };
 
-static char *gw_comm_map[] = {
+static const char *gw_comm_map[] = {
      "  ",
      "~~",
      "..",
@@ -123,7 +123,7 @@ int set_shm_debug(void)
  *
  *  returns: pointer to text version of state
  */
-char *map_aci_state(ACISTATE s) {
+const char *map_aci_state(ACISTATE s) {
      /* FIXME: this is simplistic and unchecked */
      return(aci_state_map[s]);
 }
@@ -136,7 +136,7 @@ char *map_aci_state(ACISTATE s) {
  *
  *  returns: pointer to text version of state
  */
-char *map_gw_state(GWSTATE s) {
+const char *map_gw_state(GWSTATE s) {
      /* FIXME: this is simplistic and unchecked */
      return(gw_state_map[s]);
 }
@@ -149,7 +149,7 @@ char *map_gw_state(GWSTATE s) {
  *
  *  returns: pointer to symbol string for state
  */
-char *map_gw_comm(GWSTATE s) {
+const char *map_gw_comm(GWSTATE s) {
      /* FIXME: this is simplistic and unchecked */
      return(gw_comm_map[s]);
 }

--- a/rmschanstat/rmschanstat.in
+++ b/rmschanstat/rmschanstat.in
@@ -86,49 +86,23 @@ if [ "${TYPE}" = "ax25" ]; then
     fi
 
     #
-    # look for the IP address of the port
+    # is the call listening on that specific interface?
     #
-    set -o noglob
-#    IP=($(ps ax | grep attach | egrep "${NAME}[ 	]+" | egrep -o '[0-9]{1,3}.[0-9]{1,3}.[0-9]{1,3}.[0-9]{1,3}' ))
-    IP=($(ps ax | grep attach | sed -n -e "s/^.*${NAME} //p" ))
-
-# debug:
-# IP=($(cat /tmp/test123 | grep attach | egrep "${NAME}[ 	]+" | egrep -o '[0-9]{1,3}.[0-9]{1,3}.[0-9]{1,3}.[0-9]{1,3}' ))
-# echo "IP: ${IP}"
-
-    set +o noglob
-
-    if [ -z "${IP}" ]; then
-	echo "port ${NAME} unavailable"
-	exit 1
-    fi
-
-    #
-    # get the interface using this IP address
-    # (pos 7 when using ax25, pos 8 for 6pack)
-    #
-
-
-    INTERFACE=($(ifconfig | egrep -B 2 ${IP}))
-    if [ -z "${INTERFACE}" ]; then
-        echo "interface with ${IP} unavailable"
+    INTERFACE=$(rmsfinddev ${NAME})
+    if [ $? -ne 0 ]; then
+        echo "No AX.25 device found for channel ${NAME}"
         exit 1
     fi
 
-    #
-    # is the call listening on that specific interface?
-    #
-    INTERFACE=$(echo $INTERFACE | tr -d ':')
-
     set -o noglob
-    STATUS=($(netstat --protocol=${TYPE} -l | grep "${INTERFACE[0]}" | grep "${CALL}" | grep -i "LISTENING"))
+    STATUS=($(netstat --protocol=${TYPE} -l | grep "${INTERFACE}" | grep "${CALL}" | grep -i "LISTENING"))
     set +o noglob
 
     if [ -z "${STATUS}" ]; then
-	echo "status for interface ${INTERFACE[0]} unavailable"
+	echo "status for interface ${INTERFACE} unavailable"
 	exit 1
     else
-        echo "channel ${NAME} with callsign ${CALL} on interface ${INTERFACE[0]} up"
+        echo "channel ${NAME} with callsign ${CALL} on interface ${INTERFACE} up"
         exit 0
     fi
 

--- a/rmsfinddev/Makefile.am
+++ b/rmsfinddev/Makefile.am
@@ -1,0 +1,9 @@
+bin_PROGRAMS = rmsfinddev
+rmsfinddev_SOURCES = rmsfinddev.c
+
+rmsfinddev_LDADD = -lax25
+
+AM_CPPFLAGS = -I$(top_srcdir)/include \
+               -D_GNU_SOURCE
+
+AM_CFLAGS = -Wall -O2

--- a/rmsfinddev/rmsfinddev.c
+++ b/rmsfinddev/rmsfinddev.c
@@ -1,0 +1,31 @@
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <netax25/ax25.h>
+#include <netax25/axconfig.h>
+
+int main( int argc, char *argv[])
+{
+	if( argc != 2 )
+	{
+		fprintf( stderr, "Usage: %s <portname>\n", argv[0]);
+		return -1;
+	}
+
+	if( ax25_config_load_ports() == 0 )
+	{
+		fprintf( stderr, "Failed to load the AX.25 port configuration.\n");
+		return -2;
+	}
+
+	const char *dev = ax25_config_get_dev( argv[1]);
+	if( !dev )
+	{
+		fprintf( stderr, "Unable to find device for port %s.\n", argv[1]);
+		return 1;
+	}
+
+	printf( "%s\n", dev);
+	return 0;
+}
+

